psi_cardinality/javascript/bindings/client.cpp

```diff
@@ -1,61 +1,47 @@
-}
+#include "psi_cardinality/cpp/psi_cardinality_client.h"
+#include "psi_cardinality/javascript/bindings/utils.h"
 
 EMSCRIPTEN_BINDINGS(PSI_Client) {
-    function("getException", &get_exception);
-    function("vecFromJSArray", select_overload<std::vector<std::string>(const val &)>(&vecFromJSArray));
-
-    register_vector<std::string>("std::vector<std::string>");
-        
-    class_<PSICardinalityClient>("PSICardinalityClient")
-        .smart_ptr<std::shared_ptr<PSICardinalityClient>>("std::shared_ptr<PSICardinalityClient>")
-        .class_function("Create", optional_override([]() {
-            std::shared_ptr<PSICardinalityClient> client = PSICardinalityClient::Create().ValueOrDie();
-            return client;
-        }))
-        .function("CreateRequest", optional_override([](const PSICardinalityClient &self, 
-            const std::vector<std::string> &vect) {
-            std::string request = self.CreateRequest(vect).ValueOrDie();
-            return request;
-        }))
-        .function("ProcessResponse", optional_override([](const PSICardinalityClient &self, 
-            const std::string &setup, const std::string &response) {
-            std::int64_t result = self.ProcessResponse(setup, response).ValueOrDie();
-            const std::int32_t intersection = (int32_t) result;
-            return intersection;
-        }));
+  using emscripten::optional_override;
+  using private_join_and_compute::StatusOr;
+  using psi_cardinality::PSICardinalityClient;
+  using psi_cardinality::ToJSObject;
+  using psi_cardinality::ToShared;
+
+  emscripten::class_<PSICardinalityClient>("PSICardinalityClient")
+      .smart_ptr<std::shared_ptr<PSICardinalityClient>>(
+          "std::shared_ptr<PSICardinalityClient>")
+      .class_function(
+          "Create", optional_override([]() {
+            return ToJSObject(ToShared(PSICardinalityClient::Create()));
+          }))
+      .function("CreateRequest",
+                optional_override([](const PSICardinalityClient &self,
+                                     const emscripten::val &v) {
+                  std::vector<std::string> inputs;
+                  const auto l = v["length"].as<unsigned>();
+                  inputs.reserve(l);
+
+                  for (auto i = 0; i < l; ++i) {
+                    inputs.push_back(v[i].as<std::string>());
+                  }
+
+                  return ToJSObject(self.CreateRequest(inputs));
+                }))
+      .function("ProcessResponse",
+                optional_override([](const PSICardinalityClient &self,
+                                     const std::string &setup,
+                                     const std::string &response) {
+                  // We need to convert to an int32 explicitly because JS
+                  // doesn't have 64-bit integers.
+                  StatusOr<int32_t> result;
+                  auto status = self.ProcessResponse(setup, response);
+                  if (status.ok()) {
```

Comment:
Isn't this checked in the template function `ToJSObject`

Reply:
It is, but here we're converting from a StatusOr<int64_t> to a StatusOr<int32_t>. That conversion doesn't happen implicitly. That's also the reason we need ToShared for the factory functions.